compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -794,23 +805,192 @@ generateSrcPkgFromLf config pkg = do
-    LF.TNat _ -> False
-
--- | Determine whether a typecon refers to an old-style
+-- | A reference that can appear in a type or expression. We need to track
+-- all of these in order to control what gets hidden in data-dependencies.
+data Ref
+    = RTypeCon (LF.Qualified LF.TypeConName)
+        -- ^ necessary to track hidden types and old-style typeclasses
+    | RTypeSyn (LF.Qualified LF.TypeSynName)
+        -- ^ necessary to track hidden new-style typeclasses
+    | RValue (LF.Qualified LF.ExprValName)
+        -- ^ necessary to track hidden typeclass instances
+    deriving (Eq, Ord, Show, Generic)
+
+instance Hashable Ref
+
+refPackage :: Ref -> LF.PackageRef
+refPackage = \case
+    RTypeCon q -> LF.qualPackage q
+    RTypeSyn q -> LF.qualPackage q
+    RValue q -> LF.qualPackage q
+
+refModule :: Ref -> LF.ModuleName
+refModule = \case
+    RTypeCon q -> LF.qualModule q
+    RTypeSyn q -> LF.qualModule q
+    RValue q -> LF.qualModule q
+
+type RefGraph = HMS.HashMap Ref (Bool, [Ref])
+
+-- | Calculate the set of all references that should be hidden.
+buildHiddenRefMap :: Env -> HMS.HashMap Ref Bool
+buildHiddenRefMap env =
+    case leastFixedPointBy (||) refGraphList of
+        Left ref -> error
+            ("Internal error: missing reference in RefGraph "
+            <> show ref)
+        Right m -> m
+  where
+    refGraphList :: [(Ref, Bool, [Ref])]
+    refGraphList =
+        [ (ref, hidden, deps)
+        | (ref, (hidden, deps)) <- HMS.toList refGraph
+        ]
+
+    refGraph :: RefGraph
+    refGraph = foldl visitRef HMS.empty (rootRefs env)
+
+    visitRef :: RefGraph -> Ref -> RefGraph
+    visitRef !refGraph ref
+        | HMS.member ref refGraph
+            = refGraph -- already in the map
+        | ref == RTypeCon erasedTCon
+            = HMS.insert ref (True, []) refGraph -- Erased is always erased
+        | refModule ref == LF.ModuleName ["DA", "Generics"]
+            = HMS.insert ref (True, []) refGraph -- DA.Generics is not supported
+        | LF.PRImport pkgId <- refPackage ref
+        , Set.member pkgId (configDependencyPackages (envConfig env))
+            = HMS.insert ref (False, []) refGraph -- dependencies are always available
+        | LF.PRImport pkgId <- refPackage ref
+        , MS.member pkgId (configStablePackages (envConfig env))
+            = HMS.insert ref (False, []) refGraph -- stable pkgs are always available
+
+        | RTypeCon tcon <- ref
+        , Just defDataType <- envLookupDataType tcon env
+        , refs <- DL.toList (refsFromDefDataType defDataType)
+        , hidden <- defDataTypeIsOldTypeClass defDataType
+        , refGraph' <- HMS.insert ref (hidden, refs) refGraph
+            = foldl visitRef refGraph' refs
+
+        | RTypeSyn tsyn <- ref
+        , Just defTypeSyn <- envLookupSynonym tsyn env
+        , refs <- DL.toList (refsFromDefTypeSyn defTypeSyn)
+        , refGraph' <- HMS.insert ref (False, refs) refGraph
+            = foldl visitRef refGraph' refs
```

Comment:
```suggestion
            = foldl' visitRef refGraph' refs
```
we need an HLint hint :slightly_smiling_face:

Reply:
Ahahah, yes, absolutely. :)